shared_model/backend/protobuf/common_objects/proto_common_objects_factory.hpp

```diff
@@ -0,0 +1,259 @@
+              validator_.validateAssetId(reasons, asset.assetId());
+              validator_.validateDomainId(reasons, asset.domainId());
+            });
+
+        if (answer) {
+          return iroha::expected::makeError(answer.reason());
+        }
+
+        return iroha::expected::makeValue<std::unique_ptr<interface::Asset>>(
+            std::move(proto_asset));
+      }
+
+      FactoryResult<std::unique_ptr<interface::Domain>> createDomain(
+          const interface::types::DomainIdType &domain_id,
+          const interface::types::RoleIdType &default_role) override {
+        iroha::protocol::Domain domain;
+        domain.set_domain_id(domain_id);
+        domain.set_default_role(default_role);
+
+        auto proto_domain = std::make_unique<Domain>(std::move(domain));
+
+        auto answer =
+            validate(*proto_domain, [this](const auto &domain, auto &reason) {
+              validator_.validateDomainId(reason, domain.domainId());
+              validator_.validateRoleId(reason, domain.defaultRole());
+            });
+
+        if (answer) {
+          return iroha::expected::makeError(answer.reason());
+        }
+
+        return iroha::expected::makeValue<std::unique_ptr<interface::Domain>>(
+            std::move(proto_domain));
+      }
+
+      FactoryResult<std::unique_ptr<interface::Signature>> createSignature(
+          const interface::types::PubkeyType &key,
+          const interface::Signature::SignedType &signed_data) override {
+        iroha::protocol::Signature signature;
+        signature.set_pubkey(crypto::toBinaryString(key));
+        signature.set_signature(crypto::toBinaryString(signed_data));
+
+        auto proto_singature =
+            std::make_unique<Signature>(std::move(signature));
+
+        auto answer = validate(
+            *proto_singature, [this](const auto &signature, auto &reason) {
+              validator_.validatePubkey(reason, signature.publicKey());
+            });
+
+        if (answer) {
+          return iroha::expected::makeError(answer.reason());
+        }
+
+        return iroha::expected::makeValue<
+            std::unique_ptr<interface::Signature>>(std::move(proto_singature));
+      }
+
+     private:
+      /**
+       * Perform validation of a given object
+       * @param o - object to be validated
+       * @param f - function which populates reason parameter with errors.
+       * second parameter (reasons) must be passed by non-const reference
+       * @return validation result
+       */
+      template <typename T, typename ValidationFunc>
+      validation::Answer validate(const T &o, ValidationFunc &&f) const {
+        shared_model::validation::Answer answer;
+        validation::ReasonsGroupType reasons;
+        f(o, reasons);
+        if (not reasons.second.empty()) {
+          answer.addReason(std::move(reasons));
+        }
+        return answer;
+      }
+
+      Validator validator_;
+    };
```

Comment:
There is an empty space between curly braces in the beginning of the file, but not here